### ISSUE TITLE
Some integration tests have wrong assumptions

### ISSUE THREAD

**Issue by sebgl** [elastic/cloud-on-k8s#736]
Relates #623.

In many cases, we rely on consuming a channel of requests in the integration tests, to kick off a reconciliation, and compare expected state after the reconciliation. This is kind of broken in multiple situations.

We should think about decoupling state convergence tests (they can have reconciliations run in background automatically), and reconciliation trigger tests (need to block check a reconciliation was started for a particular reason).

Example:

```
## kibana-es association controller (pseudo-code)

# delete the es cluster
1. c.delete(es)

# it should trigger a reconcile in the association controller
2. test.CheckReconcileCalled(t, requests)

# the association status should become Pending
3. test.RetryUntilSuccess(t, func() error {checkStatus(c.Get(assoc), pending)})
```

This is wrong for several reasons:

* in 2) we check that a reconcile was called following es deletion. In many cases the previous reconciliation iteration updated several resources (kibana, association, status, finalizers, etc.), enqueuing several reconciliations. We’re not sure we’re actually testing the deletion reconciliation trigger here.

* test.CheckReconcileCalled is a blocking call where we consume the channel of requests. Reconciliations do not happen automatically in background, but only once we’ve consumed the channel. Here we’re allowing exactly one reconciliation to go through. If it gets executed before the deletion is reflected in the controller k8s client cache, we’ll not have the expected result and we’ll never retry.

* In 3. we retry quite a few times, but we never allow another reconciliation to go through. Hence this is retrying over the test’s client cache inconsistencies, but not over multiple reconciliations to converge towards the desired state.

**Comment by sebgl** [comment]
One way to fix this would be to get rid of those reconciliation integration tests entirely: most of them focus on a subset of what the reconcile function actually does, require mocking a lot of stuff (es client, process manager client, secrets, etc.), and are implemented inconsistently.

Instead, we could improve our existing e2e tests so they cover what we currently test in the integration tests.
One downside is the time required to run those e2e tests compared to what we're used to with the integration tests, since we would spin up real pods here on a real k8s cluster. I'm currently investigating the cost in time.

**Comment by sebgl** [comment]
## ES reconciliation integration test vs. e2e test

I'm trying to compare what we do right now in Elasticsearch reconciliation integration test (fake resources, requires mocking, no es node running) vs. the same checks in an E2E test scenario.
Note this test is generally not enough to test the reconciliation, a lot more things happen in there and should be tested. 

### IT version (local etcd & apiserver, no pod scheduled):

* Create a 3-nodes cluster
* Check that pos are created
* Check that services are created
* Mock an observer with a fake ES client
* Check ES status eventually reports a green health
* Delete a pod
* Expect it to be recreated
* Delete services
* Expect them to be recreated

**Duration: 7 sec**

### E2E version (real k8s cluster & ES pods)

* Create a 3 nodes cluster
* Check that pos are created
* Check that services are created
* Check that cluster health is green using an es Client
* Check ES status eventually reports a green health
* Delete a pod & delete services
* Expect them to be recreated, ready, and cluster health to be green again
* Delete the stack and wait for deletion to be done

**Duration: 200 sec (~3.5 minutes)**

<details><summary>E2E tests time details</summary>
<pre>
--- PASS: TestESReconciliation (196.24s)
    --- PASS: TestESReconciliation/K8S_should_be_accessible (0.12s)
    --- PASS: TestESReconciliation/Stack_CRDs_should_exist (0.08s)
    --- PASS: TestESReconciliation/Remove_the_stack_if_it_already_exists (0.15s)
    --- PASS: TestESReconciliation/Creating_a_stack_should_succeed (0.09s)
    --- PASS: TestESReconciliation/Stack_should_be_created (0.06s)
    --- PASS: TestESReconciliation/ES_pods_should_eventually_be_running (0.04s)
    --- PASS: TestESReconciliation/Services_should_be_created (3.23s)
    --- PASS: TestESReconciliation/Pods_should_eventually_be_ready (64.67s)
    --- PASS: TestESReconciliation/Pods_should_eventually_have_the_expected_resources (0.05s)
    --- PASS: TestESReconciliation/Services_should_have_endpoints (0.08s)
    --- PASS: TestESReconciliation/Cluster_health_should_eventually_be_green (6.12s)
    --- PASS: TestESReconciliation/Every_secret_should_be_set_so_that_we_can_build_an_ES_client (0.07s)
    --- PASS: TestESReconciliation/Elasticsearch_endpoint_should_eventually_be_reachable (1.63s)
    --- PASS: TestESReconciliation/Elasticsearch_endpoint_should_eventually_be_reachable#01 (0.04s)
    --- PASS: TestESReconciliation/ES_pods_should_eventually_be_running#01 (0.05s)
    --- PASS: TestESReconciliation/Services_should_be_created#01 (0.11s)
    --- PASS: TestESReconciliation/Pods_should_eventually_be_ready#01 (0.05s)
    --- PASS: TestESReconciliation/Pods_should_eventually_have_the_expected_resources#01 (0.05s)
    --- PASS: TestESReconciliation/Services_should_have_endpoints#01 (0.12s)
    --- PASS: TestESReconciliation/Cluster_health_should_eventually_be_green#01 (0.03s)
    --- PASS: TestESReconciliation/Every_secret_should_be_set_so_that_we_can_build_an_ES_client#01 (0.07s)
    --- PASS: TestESReconciliation/Elasticsearch_endpoint_should_eventually_be_reachable#02 (1.64s)
    --- PASS: TestESReconciliation/Elasticsearch_endpoint_should_eventually_be_reachable#03 (0.04s)
    --- PASS: TestESReconciliation/Kill_a_pod (0.22s)
    --- PASS: TestESReconciliation/Wait_for_pod_to_be_deleted (3.08s)
    --- PASS: TestESReconciliation/Delete_discovery_service (0.09s)
    --- PASS: TestESReconciliation/Delete_external_service (0.12s)
    --- PASS: TestESReconciliation/ES_pods_should_eventually_be_running#02 (0.05s)
    --- PASS: TestESReconciliation/Services_should_be_created#02 (30.50s)
    --- PASS: TestESReconciliation/Pods_should_eventually_be_ready#02 (67.77s)
    --- PASS: TestESReconciliation/Pods_should_eventually_have_the_expected_resources#02 (0.06s)
    --- PASS: TestESReconciliation/Services_should_have_endpoints#02 (0.07s)
    --- PASS: TestESReconciliation/Cluster_health_should_eventually_be_green#02 (0.04s)
    --- PASS: TestESReconciliation/Every_secret_should_be_set_so_that_we_can_build_an_ES_client#02 (0.07s)
    --- PASS: TestESReconciliation/Elasticsearch_endpoint_should_eventually_be_reachable#04 (1.63s)
    --- PASS: TestESReconciliation/Elasticsearch_endpoint_should_eventually_be_reachable#05 (0.04s)
    --- PASS: TestESReconciliation/Deleting_stack_should_return_no_error (0.09s)
    --- PASS: TestESReconciliation/Stack_should_not_be_there_anymore (3.25s)
    --- PASS: TestESReconciliation/ES_pods_should_be_eventually_be_removed (9.21s)
PASS
ok      github.com/elastic/cloud-on-k8s/operators/test/e2e      198.124s
</pre>
</details>

What's taking the most time is waiting for ES pods to be ready (~1min).

**Comment by sebgl** [comment]
After discussion, we decided to:

* get rid of IT that are redundant with our current E2E tests (as part of this issue)
* run a basic e2e test that goes through most reconciliations on PR builds (issue #903)
* fix remaining IT that are not (yet) covered by E2E tests (issue #623)

**Comment by sebgl** [comment]
Fixed with #902, #904 and #906.
